starlite/_signature/field.py

```diff
@@ -0,0 +1,175 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any, Literal
+
+from typing_extensions import get_args, get_origin
+
+from starlite.constants import UNDEFINED_SENTINELS
+from starlite.params import BodyKwarg, DependencyKwarg, ParameterKwarg
+from starlite.types import Empty
+from starlite.utils import is_any, is_optional_union, is_union, make_non_optional_union
+from starlite.utils.predicates import is_generic, is_mapping, is_non_string_iterable, is_non_string_sequence
+
+__all__ = ("SignatureField",)
+
+
+@dataclass(unsafe_hash=True, frozen=True)
+class SignatureField:
+    """Abstraction representing a model field. This class is meant to replace equivalent datastructures from, other
+    libraries, for example, pydantic or msgspec.
+    """
+
+    __slots__ = (
+        "children",
+        "default_value",
+        "extra",
+        "field_type",
+        "kwarg_model",
+        "name",
+    )
+
+    children: tuple[SignatureField, ...] | None
+    """A mapping of subtypes, if any."""
+    default_value: Any
+    """Field name."""
+    extra: dict[str, Any]
+    """A mapping of extra values."""
+    field_type: Any
+    """The type of the kwarg."""
+    kwarg_model: ParameterKwarg | BodyKwarg | DependencyKwarg | None
+    """Kwarg Parameter."""
+    name: str
+    """Field name."""
```

Comment:
Can we add `parsed_parameter: ParsedParameter` to this model?

There are times when the signature modelling overwrites the parsed type for a parameter, e.g.:

```py
            elif parameter.should_skip_validation:
                attribute = attr.attrib(type=Any)
```
and then the `SignatureField` gets created with the type from the signature model, e.g.,

```py
    @classmethod
    def populate_signature_fields(cls) -> None:
        cls.fields = {
            k: SignatureField.create(
                field_type=attribute.type,
                name=k,
                default_value=attribute.default if attribute.default is not attr.NOTHING else Empty,
                kwarg_model=attribute.metadata.get("kwargs_model", None) if attribute.metadata else None,
                extra=attribute.metadata or None,
            )
            for k, attribute in attrs.fields_dict(cls).items()
        }
```

I've taken to using that `ParsedParameter` type as the source of truth of what was _actually_ typed on the handler - which is useful info to have around.

Reply:
You can do this in a follow-up

Reply:
Should the parsed parameters belong to the handler/dependency then?

Reply:
You can defintely do it in a followup - this is out of scope here since its not used.